### PR TITLE
Implemented post creation usecase

### PR DESCRIPTION
**Pull request by Ludosss**
Created usecase specific classes:

- PostCreationController
- PostCreationCommand
- PostCreationService
- PostCreationResponse

Created necessary errors for usecase:

- EmptyExerciseListException
- ExerciseInListNotFoundException
- UserNotFoundException

Created necessary database checks (and implemented in repository)

- CheckUserExistsPort

### CODE REVIEW

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package ca.utoronto.fitbook.adapter.persistence.localmemory;
 
 import ca.utoronto.fitbook.adapter.persistence.GenericRepository;
-import ca.utoronto.fitbook.application.port.in.EntityNotFoundException;
+import ca.utoronto.fitbook.application.port.in.exception.EntityNotFoundException;
 import ca.utoronto.fitbook.entity.Exercise;
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java:6]: Remove Space

File: src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package ca.utoronto.fitbook.application.port.in;
+
+public interface CheckUserExistsPort {
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java:4]: Remove Space

File: src/main/java/ca/utoronto/fitbook/application/port/in/LoadPostListPort.java [hunk 3]
```diff
@@ -1,5 +1,6 @@
 package ca.utoronto.fitbook.application.port.in;
 
+import ca.utoronto.fitbook.application.port.in.exception.EntityNotFoundException;
 import ca.utoronto.fitbook.entity.Post;
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/port/in/LoadPostListPort.java:5]: Remove Space

File: src/main/java/ca/utoronto/fitbook/application/port/in/command/PostCreationCommand.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package ca.utoronto.fitbook.application.port.in.command;
+
+import lombok.NonNull;
+import lombok.Value;
+
+import java.util.List;
+
+@Value
+public class PostCreationCommand {
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/port/in/command/PostCreationCommand.java:10]: Remove spaces

File: src/main/java/ca/utoronto/fitbook/application/port/in/exception/UserNotFoundException.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package ca.utoronto.fitbook.application.port.in.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.INTERNAL_SERVER_ERROR, reason="User not found")
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/port/in/exception/UserNotFoundException.java:6]: Not an Internal Server Error

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import ca.utoronto.fitbook.application.port.in.command.PostCreationCommand;
+import ca.utoronto.fitbook.application.port.in.exception.EmptyListException;
+import ca.utoronto.fitbook.application.port.in.exception.UserNotFoundException;
+import ca.utoronto.fitbook.application.port.out.SavePostPort;
+import ca.utoronto.fitbook.application.port.out.response.PostCreationResponse;
+import ca.utoronto.fitbook.entity.Post;
+import ca.utoronto.fitbook.entity.User;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:13]: Remove Spaces

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Date;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class PostCreationService implements PostCreationUseCase{
+    private final SavePostPort savePostPort;
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:21]: Remove Spaces

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public PostCreationResponse createPost(PostCreationCommand command) {
+
+        // Certify user existence
+        String userId = command.getUserID();
+        if (!checkUserExistsPort.checkExists(userId)) {
+            throw new UserNotFoundException("User does not exist");
+        }
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:34]: Extra Spaces should be removed

File: src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+@RequiredArgsConstructor
+public class PostCreationController {
+
+    private final PostCreationUseCase interactor;
+
+    @PostMapping(path = "/post")
+    String createPost(Model model, PostCreationCommand command){
+        PostCreationResponse outputData = interactor.createPost(command);
+        model.addAttribute("id", outputData.getPostId());
+
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java:21]: Remove Space

**Review by modaserMoj — CHANGES_REQUESTED** [review]
I highlighted some changes that in regards to your structure and ResponseStatus

File: src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java (hunk 9 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java:21]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java (hunk 8 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:34]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java (hunk 7 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:21]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java (hunk 6 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:13]: updated changes

File: src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java (hunk 9 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java:21]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/port/in/exception/UserNotFoundException.java (hunk 5 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/exception/UserNotFoundException.java]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/port/in/command/PostCreationCommand.java (hunk 4 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/command/PostCreationCommand.java]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/port/in/LoadPostListPort.java (hunk 3 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/LoadPostListPort.java:5]: updated changes

File: src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java (hunk 2 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java]: updated changes

File: src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java (hunk 1 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/adapter/persistence/localmemory/ExerciseLocalMemoryRepository.java:6]: updated changes

**Review by modaserMoj — DISMISSED** [review]
The asked changes were reviewed and implemented

File: src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.PostMapping;
+
+@Controller
+@RequiredArgsConstructor
+public class PostCreationController {
+
+    private final PostCreationUseCase interactor;
```
Comment (JacketPants12) [inline src/main/java/ca/utoronto/fitbook/adapter/web/PostCreationController.java:15]:
```suggestion
    private final PostCreationUseCase postCreationUseCase;
```
rename for clarity

File: src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package ca.utoronto.fitbook.application.port.in;
+
+public interface CheckUserExistsPort {
+    boolean checkExists(String userId);
```
Comment (JacketPants12) [inline src/main/java/ca/utoronto/fitbook/application/port/in/CheckUserExistsPort.java:4]:
```suggestion
    boolean checkUserExists(String userId);
```
rename for clarity

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+                .build();
+
+        // Save post to database
+        savePostPort.savePost(newPost);
+
+        // Fetch user from database using UserID
+        User user = loadUserPort.loadUser(userId);
+
+        // Add post to user post list
+        user.getPostIdList().add(newPost.getId());
```
Comment (JacketPants12) [inline src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:60]: Shouldn't it also save user to DB?

**Review by JacketPants12 — CHANGES_REQUESTED** [review]
Mostly changes for clarity. One comment must be addressed.

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java (hunk 12 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:62]: yup, fixed

**Review by JacketPants12 — DISMISSED** [review]
Appropriate implementation of use case

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    private final LoadUserPort loadUserPort;
+    private final CheckUserExistsPort checkUserExistsPort;
+    private final LoadExerciseListPort loadExerciseListPort;
+
+    @Override
+    public PostCreationResponse createPost(PostCreationCommand command) {
+        // Certify user existence
+        String userId = command.getUserId();
+        if (!checkUserExistsPort.checkUserExists(userId))
+            throw new UserNotFoundException("User does not exist");
```
Comment (jpablo2002) [inline src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:30]: I could be misunderstanding the code, but I think the userId should be passed when constructing and throwing the UserNotFoundException
```suggestion
            throw new UserNotFoundException(userId);
```

**Review by jpablo2002 — CHANGES_REQUESTED** [review]

**Review by jpablo2002 — APPROVED** [review]
I think everything looks good! Good job

**Review by modaserMoj — DISMISSED** [review]
Great effort!

**Review by youssefsoli — CHANGES_REQUESTED** [review]
Everything looks good, just some small naming changes are required.

File: src/main/java/ca/utoronto/fitbook/application/port/in/EmptyListException.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.UNPROCESSABLE_ENTITY, reason="Exercise list empty")
+public class EmptyListException extends RuntimeException{
+    public EmptyListException() {
+        super("Exercise list cannot be empty");
+    }
+}
```
[the remark this answers is not in the document]
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/EmptyListException.java:11]: changed, thanks

File: src/main/java/ca/utoronto/fitbook/application/port/in/ExerciseInListNotFoundException.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package ca.utoronto.fitbook.application.port.in;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value= HttpStatus.UNPROCESSABLE_ENTITY, reason="Entity not found")
```
[the remark this answers is not in the document]
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/ExerciseInListNotFoundException.java]: fixed

**Review by jpablo2002 — APPROVED** [review]
Great use of custom exceptions to provide useful error messages! Also good adherence to Clean Architecture and clean code with comments.

**Review by jpablo2002 — CHANGES_REQUESTED** [review]
After further review, I suggest refactoring your use case to use sessions  for post creation.

Here's a sample PR: #46

**Review by JacketPants12 — CHANGES_REQUESTED** [review]
Just some minor changes are needed mostly styling issues with one business logic change.

File: src/main/java/ca/utoronto/fitbook/application/port/in/EmptyExerciseListException.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package ca.utoronto.fitbook.application.port.in;
```
[the remark this answers is not in the document]
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/EmptyExerciseListException.java:1]: moved

File: src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+        // Certify exercise existence
+        List<String> exerciseIdList = command.getExerciseIdList();
+        try {
+            loadExerciseListPort.loadExerciseList(exerciseIdList);
+        } catch (EntityNotFoundException e) {
+            throw new ExerciseInListNotFoundException();
+        }
+
+        String description = command.getDescription();
```
[the remark this answers is not in the document]
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/service/PostCreationService.java:41]: added functionality and exception to limit description length to 100 characters

**Review by modaserMoj — DISMISSED** [review]
The unit testing is correct, it follows the guidelines and it tests all the code 100% coverage.

**Review by jpablo2002 — DISMISSED** [review]
I noticed that in UserFirebaseRepository for the loadUser method, you catch the EntityNotFoundException to throw a custom exception. However, in the UserLocalMemoryRepository this is not done, but I think it would be helpful to do so and be consistent.

**Review by modaserMoj — DISMISSED** [review]
The integration Testing is valid

File: src/test/java/ca/utoronto/fitbook/integration/PostCreationControllerIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+
+        MvcResult result = this.mockMvc.perform(post("/post")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(json)
+                        .session(authorizedSession))
+                .andExpect(status().isOk())
+                .andExpect(model().attributeExists("id"))
+                .andReturn();
+
+        postFirebaseRepository.loadPost((String) result.getModelAndView().getModel().get("id"));
```
Comment (OrbitalVitamin) [inline src/test/java/ca/utoronto/fitbook/integration/PostCreationControllerIntegrationTest.java:100]: What's the point of loading a post here?

**Review by OrbitalVitamin — DISMISSED** [review]
One clarification needed in integration tests, otherwise lgtm.

**Review by jpablo2002 — CHANGES_REQUESTED** [review]
It's looking fantastic, well done! There's just one last change to get the checks to pass! There's a test in the UserProfileUseCaseTest file that expects an EntityNotFoundException to be thrown. However, with the changes made, it should now be throwing the more specific and informative UserNotFoundException. Aside from that looks great!

File: src/test/java/ca/utoronto/fitbook/integration/PostCreationControllerIntegrationTest.java (hunk 18 above)
Comment (Ludosss) [reply src/test/java/ca/utoronto/fitbook/integration/PostCreationControllerIntegrationTest.java:100]: loading a post to ensure that the newly created post exists in the database with the returned id

**Review by modaserMoj — APPROVED** [review]
Thank you for applying all the feedback, I think this is  good to go.

File: src/main/java/ca/utoronto/fitbook/application/port/in/UserNotFoundException.java [hunk 19]
```diff
@@ -0,0 +1,11 @@
+package ca.utoronto.fitbook.application.port.in;
```
Comment (jpablo2002) [inline src/main/java/ca/utoronto/fitbook/application/port/in/UserNotFoundException.java:1]: This should be in the exceptions folder

**Review by jpablo2002 — CHANGES_REQUESTED** [review]

File: src/main/java/ca/utoronto/fitbook/application/port/in/UserNotFoundException.java (hunk 19 above)
Comment (Ludosss) [reply src/main/java/ca/utoronto/fitbook/application/port/in/UserNotFoundException.java:1]: - thanks, changes made

File: src/main/resources/templates/postCreationSuccess.html [hunk 20]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Comment (jpablo2002) [inline src/main/resources/templates/postCreationSuccess.html:1]: Just a quick change. The following [Profile Use Case PR](https://github.com/CSC207-2022F-UofT/course-project-fitbook/pull/46/files#top) has the updated profile.html page with a nav section and respective styling. 

Could you just copy that appropriately into this html file so the user can leave the success page?

**Review by jpablo2002 — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/postCreationSuccess.html (hunk 20 above)
Comment (Ludosss) [reply src/main/resources/templates/postCreationSuccess.html:1]: got it, file being updated

**Review by youssefsoli — APPROVED** [review]
Nice everything checks out, we finally get to merge this beast in. Congratulations.

**Review by jpablo2002 — APPROVED** [review]
Great job implementing this vital use case following clean architecture nicely! The code is nicely documented with testing that considers various cases, and custom exceptions with descriptive error messages! Well done!

**Pull request merged**